src/stirling/protocol_inference/data_collection.py

```diff
@@ -16,32 +16,44 @@
 
 import os
 import subprocess
-import json
 import argparse
 
 
-def create_pod2pid():
+def create_pod2pid(runtime):
     """
-    Build a mapping of running pod names to host pids.
+    Builds a mapping of running pod names to host pids.
     """
-    # This cmd gets the host pid of each running container and its corresponding labels.
-    list_ps_cmd = '''kubectl node-shell ''' + node + ''' -- bash -c "docker ps -q | \
-    xargs docker inspect --format '{{.State.Pid}}{{json .Config.Labels}}'"'''
+
+    # This command prints one line of output per container where each line is the pid of the container, a
+    # single space and the name of the k8s pod determined by the io.kubernetes.pod.name container label.
+    # See the output below for example output:
+
+    # $ sudo crictl ps -q | sudo xargs crictl inspect --template \
+    #     '{{.info.pid}} {{index .info.config.labels "io.kubernetes.pod.name"}}' -o go-template
+    # 3635260 vizier-pem-sx7pr
+    # 3634678 kelvin-cdf78c57c-qlzlb
+
+    list_ps_cmd = '''kubectl node-shell ''' + node + ''' -- sudo bash -c "crictl ps -q | \
+    xargs crictl inspect -o go-template \
+    --template '{{.info.pid}} {{index .info.config.labels \\"io.kubernetes.pod.name\\"}}'"'''
+    if runtime == 'docker':
+        list_ps_cmd = '''kubectl node-shell ''' + node + ''' -- sudo bash -c "docker ps -q | \
+    xargs docker inspect --format '{{.State.Pid}} {{index .Config.Labels \\"io.kubernetes.pod.name\\"}}'"'''
```

Comment:
For this context, please consider the following construct:
```
ps_cmds = {
  'docker': docker_ps_cmd,
  'crio': crio_ps_cmd,
}
ps_cmd = ps_cmds[runtime]
```

Since this is not prod. code, this is a concise way to error check things, i.e. it reduces the possibility of picking up something wrong as new code is introduced. Granted this is not a place where we have a lot of change and churn, so maybe defaults and if/else is not as big of a hazard here. For non-perf critical prod code, I would still endorse the above scheme, but would craft things such that a missing map entry does not cause program termination.

Reply:
Updated to use an approach in spirit with your suggestion. I added a bit more templating since I felt it made dealing with the various arguments and quoting more readable.

Let me know what you think.